### PR TITLE
Improve events heap allocations

### PR DESCRIPTION
**Pull request by ValarDragon**
## Description

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: types/events.go [hunk 1]
```diff
@@ -178,7 +191,7 @@ func (a Attribute) String() string {
 
 // ToKVPair converts an Attribute object into a Tendermint key/value pair.
 func (a Attribute) ToKVPair() abci.EventAttribute {
-	return abci.EventAttribute{Key: toBytes(a.Key), Value: toBytes(a.Value)}
+	return abci.EventAttribute{Key: []byte(a.Key), Value: []byte(a.Value)}
```
Comment (ValarDragon) [inline types/events.go:194]: This had a signficant impact on performance!

However its still silly that we have to keep re-allocating heap space for the keys, when ~all the time, they are actually constants. 

I think we should change the data-backend here to be using `[]byte`, at least until abci events use strings.

File: x/bank/keeper/send.go [hunk 2]
```diff
@@ -138,18 +138,16 @@ func (k BaseSendKeeper) InputOutputCoins(ctx sdk.Context, inputs []types.Input,
...
-			sdk.NewAttribute(types.AttributeKeySender, fromAddr.String()),
-		),
-	})
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeTransfer,
+		sdk.NewAttribute(types.AttributeKeyRecipient, toAddr.String()),
+		sdk.NewAttribute(types.AttributeKeySender, fromAddr.String()),
+		sdk.NewAttribute(sdk.AttributeKeyAmount, amt.String()),
+	))
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
```
Comment (ValarDragon) [inline x/bank/keeper/send.go:147]: Kind of annoying, but this also actually improved performance of the entire benchmark by a couple percent, and moreso the event sub-set of it. It avoids extra heap allocations on the argument, and then avoids a slice copy, instead you just append a stack element.

File: types/events.go [hunk 3]
```diff
@@ -210,15 +223,9 @@ func (e Events) ToABCIEvents() []abci.Event {
 	return res
 }
 
-func toBytes(i interface{}) []byte {
-	switch x := i.(type) {
-	case []uint8:
```
Comment (UnityChaos) [inline types/events.go:215]: I assume this case is never actually used?

File: x/bank/keeper/send.go (hunk 2 above)
Comment (UnityChaos) [reply x/bank/keeper/send.go:147]: Hmm is the extra `EventTypeMessage` actually used anywhere? I assume that's a general event emitted on basically all transactions that have a sender. It definitely seems like it makes more sense to just have the more specific `EventTypeTransfer` here, but are we sure no one is indexing / searching by the more general type?

Comment (UnityChaos) [reply x/bank/keeper/send.go:147]: Oh wait, you're not actually getting rid of it lol, just separating it into two EmitEvents. The diff got cut off lol.

**Review by UnityChaos — APPROVED** [review]
LGTM :+1:

File: types/events.go (hunk 3 above)
Comment (ValarDragon) [reply types/events.go:215]: Yeah, and it was bad to be casting to interface and then reflecting.

File: x/bank/keeper/send.go (hunk 2 above)
Comment (ValarDragon) [reply x/bank/keeper/send.go:147]: ah woops, I should've copied more of the diff in the comment

**Pull request merged**